R/stan_module.R

```diff
@@ -0,0 +1,126 @@
+#' StanAll class object
+#' @slot functions Character, the functions part of a stan model.
+#' @slot data Character, the data part of a stan model.
+#' @slot parameters Character, the parameters part of a stan model.
+#' @slot transformed_parameters Character, the transformed_parameters part of a stan model.
+#' @slot model Character, the model part of a stan model.
+#' @slot priors List, the priors part of a stan model.
+#' @slot generated_quantities Character, the generated_quantities part of a stan model.
+#' @slot includes Character
+#' @slot inits List with the initial values of the stan model.
+#' @exportClass StanModule
+StanModule <- setClass(
+    "StanModule",
+    representation(
+        functions = "character",
+        data = "character",
+        parameters = "character",
+        transformed_parameters = "character",
+        priors = "list",
+        generated_quantities = "character",
+        inits = "list"
+    )
+)
+
+
+
+#' read_stan_part returns stan code as a character.
+#' @param file Character, either the absolute path of a stan file, or the name of the stan file in the package directory or the stan code as a string.
+#' @export
+read_stan_part <- function(file) {
+    if (file.exists(file)) {
+        absolute_filename <- file
+        out <- readLines(absolute_filename)
+    } else if (file.exists(system.file("stanparts", file, package = "jmpost"))) {
+        absolute_filename <- system.file(
+            "stanparts",
+            file,
+            package = "jmpost"
+        )
+
+        out <- readLines(absolute_filename)
+    } else {
+        out <- file
+    }
+
+    return(out)
+}
+
+
+#' @importFrom assertthat assert_that
+#' @param functions TODO
+#' @param data TODO
+#' @param parameters TODO
+#' @param transformed_parameters TODO
+#' @param priors TODO
+#' @param generated_quantities TODO
+#' @param inits TODO
+#' @export
+setMethod(
+    f = "initialize",
+    signature = "StanModule",
+    definition = function(.Object, ..., functions, data,
+                          parameters, transformed_parameters,
+                          priors, generated_quantities, inits) {
+        assert_that(
+            is.character(functions),
+            is.character(data),
+            is.character(parameters),
+            is.character(transformed_parameters),
+            is.character(generated_quantities),
+            msg = "`Functions`, `data`, `parameters`, `transformed_parameters` and `generated_quantities` must be character vectors"
+        )
+
+        assert_that(
+            is.list(priors),
+            is.list(inits),
+            length(inits) == length(priors),
```

Comment:
I think we need to allow for `priors` being populated and `inits` being left blank. This currently forces you to define your own `init` even if you'd rather leave it up to Stan to do that for you

Reply:
The idea is that the default `inits` will be a list of either `NAs`, values or `functions`. This list will be curried on until the moment when the data are merged to the same object as the code. Then, the functions within the `inits` list will generate the necessary initial values.  For example the default list of the initial values of the os model https://github.com/Genentech/jmpost/blob/c9b971708856990559d88e31026a2402b1bf0fd9/R/inits.R#L25-L29

Reply:
I think it would be good to assign sensible defaults here  to avoid users running into this error which might be quite annoying, perhaps something like:

```
    definition = function(
        .Object,
        <other arguments>
        priors = list(),
        inits =  rep(NA, length(priors))
    )
```

Reply:
>         inits =  rep(NA, length(priors))

The issue here is that the inits list might contain an element that is a list or a vector of length defined from the data object. Thus, the stan code will expect eg `inits$beta_os_cov` to be a vector of length 5 and it will get only one `NA`.

Reply:
Fair point, maybe instead we should just support `inits` being a single `NA` value which we can then replicate across later when we reduce the code

Reply:
What about being a list of either `NA` values or functions? I do believe that there is a benefit of carrying this list of values or functions from the beginning. This way, the user can modify this list at any point.

Reply:
O yes, sorry I'm not saying not support functions / list of values, what I'm trying to implement is a simple way for the user to say "I don't want to set priors for anything"  which I think just providing a single NA value (or a null or a blank list) would be quite clean.

Thinking about it I think a blank list is fine,  I  am thinking that if any parameter is missing from inits then we should assume they dont want to specify a starting value for that parameter (assuming stan can support this)

Reply:
I see your point. Lets allow the user to provide an empty list at this stage.